tests/candev/main.c

```diff
@@ -0,0 +1,213 @@
+        isrpipe_read(&rxbuf, buf, 1);       //dlc
+        can_dlc = buf[0];
+        isrpipe_read(&rxbuf, buf, can_dlc); //data
+
+        printf("id: %" PRIx32 " dlc: %" PRIx8 " Data: \n", can_id, can_dlc);
+        for (int i = 0; i < can_dlc; i++) {
+            printf("0x%X ", buf[i]);
+        }
+        puts("");
+    }
+
+    return 0;
+}
+
+static const shell_command_t shell_commands[] = {
+    { "send", "send some data", _send },
+    { "receive", "receive some data", _receive },
+    { NULL, NULL, NULL }
+};
+
+static void _can_event_callback(candev_t *dev, candev_event_t event, void *arg)
+{
+    (void)arg;
+    struct can_frame *frame;
+
+    switch (event) {
+        case CANDEV_EVENT_ISR:
+            DEBUG("_can_event: CANDEV_EVENT_ISR\n");
+            dev->driver->isr(candev);
+            break;
+        case CANDEV_EVENT_WAKE_UP:
+            DEBUG("_can_event: CANDEV_EVENT_WAKE_UP\n");
+            break;
+        case CANDEV_EVENT_TX_CONFIRMATION:
+            DEBUG("_can_event: CANDEV_EVENT_TX_CONFIRMATION\n");
+            break;
+        case CANDEV_EVENT_TX_ERROR:
+            DEBUG("_can_event: CANDEV_EVENT_TX_ERROR\n");
+            break;
+        case CANDEV_EVENT_RX_INDICATION:
+            DEBUG("_can_event: CANDEV_EVENT_RX_INDICATION\n");
+
+            frame = (struct can_frame *)arg;
+
+            DEBUG("\tid: %" PRIx32 " dlc: %" PRIx8 " Data: \n\t", frame->can_id,
+                  frame->can_dlc);
+            for (uint8_t i = 0; i < frame->can_dlc; i++) {
+                DEBUG("0x%X ", frame->data[i]);
+            }
+            DEBUG(" ");
+
+            //Store in buffer until user requests the data
+            isrpipe_write_one(&rxbuf,
+                              (uint8_t)((frame->can_id & 0x1FFFFFFF) >> 24));        //exclude flags
+            isrpipe_write_one(&rxbuf,
+                              (uint8_t)((frame->can_id & 0xFF0000) >> 16));
+            isrpipe_write_one(&rxbuf, (uint8_t)((frame->can_id & 0xFF00) >> 8));
+            isrpipe_write_one(&rxbuf, (uint8_t)((frame->can_id & 0xFF)));
+
+            isrpipe_write_one(&rxbuf, frame->can_dlc);
+            for (uint8_t i = 0; i < frame->can_dlc; i++) {
+                isrpipe_write_one(&rxbuf, frame->data[i]);
+            }
+
+            break;
+        case CANDEV_EVENT_RX_ERROR:
+            DEBUG("_can_event: CANDEV_EVENT_RX_ERROR\n");
+            break;
+        case CANDEV_EVENT_BUS_OFF:
+            dev->state = CAN_STATE_BUS_OFF;
+            break;
+        case CANDEV_EVENT_ERROR_PASSIVE:
+            dev->state = CAN_STATE_ERROR_PASSIVE;
+            break;
+        case CANDEV_EVENT_ERROR_WARNING:
+            dev->state = CAN_STATE_ERROR_WARNING;
+            break;
+        default:
+            DEBUG("_can_event: unknown event\n");
+            break;
+    }
+}
+
+int main(void)
+{
+    uint8_t rx_ringbuf[RX_RINGBUFFER_SIZE] = { 0 };
```

Comment:
You don't want to put that in the main stack - just move it outside the main function.

Reply:
I moved it outside, but could you elaborate on why it's best to not put that in main? (Because on AVR, bss is very limited. Every byte counts :) )

Reply:
It's in the RAM no matter whether you put it in the thread stack or on global memory.
But in the thread stack it's 'hidden' - if you increase `RX_RINGBUFFER_SIZE` you might overflew the main thread stack without even knowing it.
And that's a failure at run-time.

If you need to save RAM, you might as well decrease your main stack size by 128 bytes now.
But beware, it's already only 512 bytes on AVR. (Or 384 bytes with that buffer in it)

Reply:
Ok, thanks for the explanation!
I'll leave the buffersize unchanged for now, because one can message is already up to 16 bytes. Halving it would only allow for 4 messages to be stored in the buffer. 
If someone wants to use the app in a memory-constrained environment they can still lower the buffersize :)
(Also because currently only native is supported, and we usually don't have much memory problems on native ;) )